### ISSUE TITLE
E0392 incorrectly claims generic parameter used in associated type bound is unused

### ISSUE THREAD

**Issue by PatchMixolydic** [rust-lang/rust#105311]
Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8156598825422fce2a021610c9ae54a0)):
```rust
trait Trait<T> {
    type Type;
}

struct X<U, F: Trait<(), Type = U>>(F);
struct Y<T, U, F: Trait<T, Type = U>>(F);
```

The current output is:
```
error[E0392]: parameter `T` is never used
 --> src/lib.rs:6:10
  |
6 | struct Y<T, U, F: Trait<T, Type = U>>(F);
  |          ^ unused parameter
  |
  = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
  = help: if you intended `T` to be a const parameter, use `const T: usize` instead

error[E0392]: parameter `U` is never used
 --> src/lib.rs:6:13
  |
6 | struct Y<T, U, F: Trait<T, Type = U>>(F);
  |             ^ unused parameter
  |
  = help: consider removing `U`, referring to it in a field, or using a marker such as `PhantomData`
  = help: if you intended `U` to be a const parameter, use `const U: usize` instead
```

Note that there are no diagnostics for `X`. Indeed, removing `Y` causes the build to succeed.

The first instance of E0392 is allegedly correct, per #60214 and #53589. However, the second diagnostic claiming that `U` is unused seems to be incorrect. Otherwise, E0392 should also complain about `X`'s `U` parameter.

Ideally, the output should look like:
```
error[E0392]: parameter `T` is never used
 --> src/lib.rs:6:10
  |
6 | struct Y<T, U, F: Trait<T, Type = U>>(F);
  |          ^ unused parameter
  |
  = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
  = help: if you intended `T` to be a const parameter, use `const T: usize` instead
```

@rustbot modify labels +A-traits +D-incorrect
